### ISSUE TITLE
Law-text-Embedding-Project

### ISSUE THREAD

**Issue by Mahdi-122** [Mahdi-122/T-122#5]
from datasets import load_dataset
from scipy.sparse import lil_matrix
import numpy as np
from sklearn.decomposition import TruncatedSVD
from sklearn.manifold import TSNE
import plotly.express as px
import pandas as pd
import re
from collections import Counter

#First we load our Dataset
LawDataset = load_dataset("QomSSLab/law-text-dataset-fa")
LawDataTxt = LawDataset["train"]["text"]

#then we normalize our data set
def Normalize():
    global NomText
    NomText = []
    for txt in LawDataTxt :
        txt = txt.replace("ي","ی")
        txt = txt.replace("ك","ک")
        txt = txt.replace("ة","ه")
        txt = txt.replace("ؤ","و")
        txt = txt.replace("ئ","ی")
        txt = txt.replace("أ","ا")
        txt = txt.replace("إ","ا")
        txt = txt.replace("آ","ا")
        txt = txt.replace("-","")
        txt = txt.replace("،", "")
        txt = txt.replace("\n"," ")
        txt = txt.replace("\u200c"," ")
        txt = txt.replace("\u200f", "")
        txt = txt.replace("\u200e", "")
        txt = txt.replace("–", "")
        txt = txt.replace("\u200b", "")
        txt = txt.replace("\u200d", "")
        txt = txt.replace("►", "")
        txt = txt.replace("═", "")
        txt = txt.replace("≤", "")
        txt = txt.replace(".", "")
        txt = txt.replace("»", "")
        txt = txt.replace("«", "")
        txt = txt.replace("**","")
        txt = txt.replace("\"","")
        txt = txt.replace("۰","")
        txt = txt.replace("۱","")
        txt = txt.replace("۲","")
        txt = txt.replace("۳","")
        txt = txt.replace("۴","")
        txt = txt.replace("۵","")
        txt = txt.replace("۶","")
        txt = txt.replace("۷","")
        txt = txt.replace("۸","")
        txt = txt.replace("۹","")
        txt = txt.replace("#","")
        txt = txt.replace("&","")
        txt = txt.replace("…","")
        txt = txt.replace(")","")
        txt = txt.replace("(","")
        txt = txt.replace("!","")
        txt = txt.replace(">","")
        txt = txt.replace("<","")
        txt = re.sub(r'http[s]?://\S+|www\.\S+', '', txt)
        txt = txt.strip()

        NomText.append(txt)
    return NomText

Normalize()

#after normalizing our data set we start tokenizing
def Tokenizer():
    Tokens = []
    for t in NomText:
        Tokens.append(t.split())
    return Tokens

Tokenized = Tokenizer()

#building our vocabulary 
all_words = [w for s in Tokenized for w in s]
unique_w = sorted(list(set(all_words)))
UW_size = len(unique_w)
#and index our words and index back to words for checking back on it

Word2index = {word: i for i, word in enumerate(unique_w)}
index2Word = {i: word for word, i in Word2index.items()}

#we ask for the window size if it should be 1 or 2
window_size = int(input("what Window size should we use (1/2) : "))

#are cooccurrence sparse matrix
CoMat = lil_matrix((UW_size, UW_size), dtype=np.int32)

for s in Tokenized:
    inc = [Word2index[w] for w in s if w in Word2index]
    for l, center in enumerate(inc):
        start = max(0, l - window_size)
        end = min(len(inc), l + window_size + 1)
        for k in range(start, end):
            if l != k:
                CoMat[center, inc[k]] += 1

Dense_CM = CoMat.toarray().astype(np.float32)
MeanC = Dense_CM - Dense_CM.mean(axis=0)

compactv = TruncatedSVD(n_components=100)
WordE = compactv.fit_transform(MeanC)

#check our word similarities 
def word_similarity(comp_w1, comp_w2):
    if comp_w1 not in Word2index:
        return f"Unknown word: {comp_w1}"
    if comp_w2 not in Word2index:
        return f"Unknown word: {comp_w2}"

    RowI1 = Word2index[comp_w1]
    RowI2 = Word2index[comp_w2]

    vector1 = WordE[RowI1]
    vector2 = WordE[RowI2]

    ham = np.dot(vector1, vector2)
    cm = np.linalg.norm(vector1) * np.linalg.norm(vector2)

    if cm == 0:
        return "whoops Invalid vector ."

    return ham / cm

wordC1 = input("Choose your what word you want to compare : ")
wordC2 = input("Choose the second your what word you want to compare : ")

print("Similarity :", word_similarity(wordC1, wordC2))

#pick 1000 most frequent words
word_frec = Counter(all_words)
mostCommonW = [w for w, c in word_frec.most_common(1000)]
tw_indices = [Word2index[w] for w in mostCommonW]
tw_embedding = WordE[tw_indices]

#
mdl = TSNE(n_components=2, perplexity=30, learning_rate=200)
embedTo2d = mdl.fit_transform(tw_embedding)

#reduce to 2d
dataFrame = pd.DataFrame({
    "x": embedTo2d[:, 0],
    "y": embedTo2d[:, 1],
    "word": mostCommonW
})
#use plotty for Visualizing 
Em_plot = px.scatter(dataFrame, x="x", y="y", text="word", width=1500, height=900)
Em_plot.update_traces(textposition='top center')
Em_plot.show()
